### PR TITLE
Kcwi

### PR DESCRIPTION
**Pull request by SunilSimha**
A new analysis module to produce MARZ files from KCWI datacubes. This requires the following packages:
`sep`: Source Extractor for Python
`PyRegion`: For handling ds9 region files
`SpectralCube`: To handle fits datacubes.

All functions can in principle, work with any datacube, not just KCWI. This has only been tested on KCWI cubes, however.
 
Includes:
* Making images from cubes (including transmission curves if necessary).
* Extracting spectra from masked cubes.
* Source detection in 2D images.
* Extracting spectra of identified sources.

### CODE REVIEW

File: frb/analysis/kcwi.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    raise ImportError("Requirement unmet: SpectralCube. Run `pip install spectral-cube`.")
+try:
+    import pyregion as pyreg
+except ImportError:
+    raise ImportError("Requirement unmet: pyregion. Run `pip install pyregion`.")
+
+import glob, os, sys
+
+def _air_to_vac(wave):
+    """
```
Comment (profxj) [inline frb/analysis/kcwi.py]: Add doc on parameter `wave`

File: frb/analysis/kcwi.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    sigma2 = (1e4/wave)**2
+    fact = 1.+5.792105e-2/(238.0185-sigma2)+1.67917e-3/(57.362-sigma2)
+    return wave*fact 
+
+def _spectral_tile(array, cube):
+    """
+    Helper function that tiles 1D array of
+    size cube.spectral_axis.shape and tiles it
+    to the same shape as the cube.
+    """
```
Comment (profxj) [inline frb/analysis/kcwi.py]: add doc on parameters `array`, `cube`

do the same for all other methods.

File: frb/analysis/kcwi.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    Flatten cube along wavelength and produce a 2D
+    image.
+    Args:
+        cubefile (str): Path to the datacube
+        wlow, whigh (Quantity, optional): wavelength 
+            limits (with astropy units) to flatten between.
+            If nothing is given, the cube is checked for the
+            WAVGOOD keywords and flattened between
+            them. If they don't exist, it's flattened fully.
+        filter (function, optional): transmission
```
Comment (profxj) [inline frb/analysis/kcwi.py]: I don't see this in the parameter list

File: frb/analysis/kcwi.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    
+    return img
+
+def spec_from_mask(cube, mask_arr, varcube=None, kind="mean"):
+    """
+    Extract a spectrum from a cube
+    within a mask.
+    Args:
+        cube (Spectral Cube): A datacube object
+        mask_arr (numpy array): A 2D boolean array
```
Comment (profxj) [inline frb/analysis/kcwi.py]: Is this mask intended to identify which pixels are extracted?
I think so.  Please specify.

File: frb/analysis/kcwi.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        spec (OneDSpectrum): The extracted spectrum.
+        var (OneDSpectrum): Variance in spectrum.
+            Only returned if varcube is supplied.
+    """
+    #TODO: Use photutils aperture object for this.
+    # Create 2D mask first
+    mask = np.zeros(cube.shape[1:], dtype=np.bool)
+    sep.mask_ellipse(mask, x0, y0, a, b, theta,r)
+    
+    return spec_from_mask(cube, mask, varcube)
```
Comment (profxj) [inline frb/analysis/kcwi.py]: Make it `varcube=varcube`

File: frb/analysis/kcwi.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+                                  minarea = minarea, clean = clean,
+                                  segmentation_map=True)
+    if write:
+        Table(objects).write(write)
+    
+    return Table(objects), segmap
+
+def _make_marz(cube, speclist, varspeclist, objects,marzfile="marzfile.fits", tovac = True):
+    """
+    Helper function to create a MARZ input file
```
Comment (profxj) [inline frb/analysis/kcwi.py]: doc str on parameters

File: frb/analysis/kcwi.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+    # Initialize output lists 
+    speclist = np.zeros([nobjs, len(wave)])
+    varspeclist = np.zeros_like(speclist)
+
+    # Create output folder?
+    if not os.path.isdir(outdir):
+        os.mkdir(outdir)
+    for idx, obj in enumerate(objects):
+        # TODO: Make a specdb file instead of saving
```
Comment (profxj) [inline frb/analysis/kcwi.py]: Individual spectra are fine.  We can make the `specdb` file after.

File: frb/analysis/kcwi.py (hunk 1 above)
Comment (SunilSimha) [reply frb/analysis/kcwi.py]: Done

File: frb/analysis/kcwi.py (hunk 5 above)
Comment (SunilSimha) [reply frb/analysis/kcwi.py]: Done

File: frb/analysis/kcwi.py (hunk 4 above)
Comment (SunilSimha) [reply frb/analysis/kcwi.py]: Yes. Changed now.

File: frb/analysis/kcwi.py (hunk 6 above)
Comment (SunilSimha) [reply frb/analysis/kcwi.py]: Added.

File: frb/analysis/kcwi.py (hunk 7 above)
Comment (SunilSimha) [reply frb/analysis/kcwi.py]: Sounds good!

**Pull request merged**